New matcher asserting $array[$key] === $expectedValue

echochamber opened the pull request:
New matcher that allows for asserting a specific value for a specific key of methods that return an array and or an implementor of ArrayAccess.

Example Spec:

```php
<?php

SomeClassSpec extends ObjectBehavior
{

    function it_returns_an_array_with_myValue_for_myKey()
    {
        $this->someMethodThatReturnsAnArray()->shouldHaveKeyWithValue('myKey', 'myValue');
    }
}
```

Example method under test
```php
<?php

SomeClass
{

    public function someMethodThatReturnsAnArray()
    {
        return ['myKey' => 'myValue'];
    }
}

```

ciaranmcnulty commented:
Can you add some feature coverage for this please?

ciaranmcnulty commented on src/PhpSpec/Matcher/ArrayKeyValueMatcher.php:
```diff
@@ -0,0 +1,104 @@
+<?php
+
```
There needs to be a docblock here with copyright info

ciaranmcnulty commented on src/PhpSpec/Matcher/ArrayKeyValueMatcher.php:
```diff
@@ -0,0 +1,104 @@
+            ));
+        }
+
+        // Match failed because key was set but the value wasn't what was expected
```
don't think you need these comments.

ciaranmcnulty commented on src/PhpSpec/Matcher/ArrayKeyValueMatcher.php:
```diff
@@ -0,0 +1,104 @@
+        $actualValue = $subject[$key];
+
+        // Match failed because the key wasn't set
+        if (($subject instanceof ArrayAccess && $subject->offsetExists($key)) || array_key_exists($key, $subject)) {
```
conditional would be clearer if extracted to a private method

ciaranmcnulty commented:
@echochamber are you going to get a chance to look at this again?

echochamber commented:
@ciaranmcnulty Hey, thanks for the ping. I totally forgot I had opened this PR. I'm busy tonight but I'll make this my priority tomorrow.

ciaranmcnulty commented:
Great thanks, if you need help with the feature coverage let me know

ciaranmcnulty commented on src/PhpSpec/Matcher/ArrayKeyValueMatcher.php:
```diff
@@ -0,0 +1,104 @@
+        $expectedValue = $arguments[1];
+        $actualValue = $subject[$key];
+
+        // Match failed because the key wasn't set
```
please remove the comment

It's better to pull bits of this out into privates so that the overall shape is clear

ciaranmcnulty commented:
We're intending to release 2.2.0-beta after the next Prophecy release so if you can get this done in the next week or so I can bundle it in - it's a good addition.

We'll also need documentation but we can raise that separately to this PR

cakper commented:
@ciaranmcnulty @echochamber by asserting key in the array you are asserting the state of the object and not it's behaviour. I don't consider this as a good object oriented design practice and I think we should not encourage people to do so by enabling such features by default.

ciaranmcnulty commented:
We assert state of return values all the time (e.g. `shouldMatch`)

ciaranmcnulty commented:
Unless you're arguing against methods that return arrays.

echochamber commented:
@ciaranmcnulty Ok, I have fixed all the issues addressed in your comments and added a feature. If there are any other changes you'd like please let me know :smile:

echochamber commented:
@ciaranmcnulty Also regarding documentation, I can write some and open a PR once this one is merged.

ciaranmcnulty commented:
The feature doesn't appear to pass

echochamber commented:
@ciaranmcnulty Fixed now.

ciaranmcnulty commented:
Thanks!
